File: 703.数据流中的第-k-大元素.go
/*
 * @lc app=leetcode.cn id=703 lang=golang
 *
 * [703] 数据流中的第 K 大元素
 */

// @lc code=start
type KthLargest struct {
	sort.IntSlice
	k int
}

func Constructor(k int, nums []int) KthLargest {
	kl := KthLargest{k: k}
	for _, num := range nums {
		kl.Add(num)
	}
	return kl
}

func (this *KthLargest) Push(x interface{}) {
	this.IntSlice = append(this.IntSlice, x.(int))
}

func (this *KthLargest) Pop() (v interface{}) {
	this.IntSlice, v = (this.IntSlice)[:this.Len()-1], (this.IntSlice)[this.Len()-1]
	return
}

func (this *KthLargest) Add(val int) int {
	heap.Push(this, val)
	if this.Len() > this.k {
		heap.Pop(this)
	}
	return this.IntSlice[0]
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * obj := Constructor(k, nums);
 * param_1 := obj.Add(val);
 */
// @lc code=end

File: 703.数据流中的第-k-大元素.py
#
# @lc app=leetcode.cn id=703 lang=python3
#
# [703] 数据流中的第 K 大元素
#

# @lc code=start
import heapq
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self._minHeap = []
        heapq.heapify(self._minHeap)
        self.k = k
        self.nums = nums
        if len(self.nums) > self.k:
            for i in range(self.k):
                heapq.heappush(self._minHeap, self.nums[i])
            for i in range(self.k, len(self.nums)):
                if nums[i] > self._minHeap[0]:
                    heapq.heappop(self._minHeap)
                    heapq.heappush(self._minHeap, self.nums[i])
        else:
            for i in range(len(self.nums)):
                heapq.heappush(self._minHeap, self.nums[i])

    def add(self, val: int) -> int:
        self.nums.append(val)
        if len(self.nums) <= self.k:
            heapq.heappush(self._minHeap, val)
        else:
            if val > self._minHeap[0]:
                heapq.heappop(self._minHeap)
                heapq.heappush(self._minHeap, val)
        return self._minHeap[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)
# @lc code=end

